### ISSUE TITLE
📊📈 Improvements to Model Infrastructure

### ISSUE THREAD

**Issue by xehu** [Watts-Lab/team_comm_tools#116]
The initial commits to `yashveer/baseline_model` are fantastic, and this Issue summarizes the next steps and open questions for improving the modeling pipeline.

To recap, the V0 baseline model currently allows the user to create a `ModelBuilder` object for each dataset individually and fit an `XGBoost` model to the dataset. Then, it identifies the most important features using Shapley values. These initial models make it possible to identify the most important features for determining success, conditional on a specific task. When we run these models, we get sensible, interpretable findings that largely align with ideas we've seen in literature reviews --- in and of itself, this is a huge step forward, and is extremely exciting!

# Overview of Next Steps
I have broken down our next steps into 3 sections:
1. Modifying the Model Pipeline to Incorporate Task Features
2. Improving Usability
3. Improving Model Optimization

In some of these steps, I made some initial code modifications to get us started, and I have marked these with "⏩️" to indicate that some progress has already been made. In the latest commit, my changes are also marked with the comment `[NEW]` so you can identify and build off of them.

## Modifying the Model Pipeline to Incorporate Task Features
The most important next step in the modeling pipeline is modifying the models to analyze similarities and differences _across tasks_. That is, how can we incorporate task features into the model, so that we can understand not only the most important conversation features conditional on some specific task, but also identify how participants change their conversational strategies in response to task demands?

Key next steps are:

- ⏩️ Modifying the pipeline to combine data from multiple datasets, rather than operating on one dataset at a time.
- ⏩️ Modifying the pipeline to incorporate task features [My "Level 0" edit incorporates each task as a dummy variable. I have more thoughts on how we can improve this, which I'll say more about below.]
- Adding in more inference options, and incorporating domain knowledge, to model how the task impacts communication, which in turn impacts performance. [This is the most open-ended component!]

### Incorporating Task Features
_Level 1: Adding the Task Map_
I have another project, called "Task Mapping," in which I create vectors for different types of team tasks. Each task occupies a unique row of the matrix, stored here: https://github.com/Watts-Lab/task-mapping/blob/master/task_map.csv

Many of the tasks used in our dataset are described in the Task Map. For example:

- Mock Jury
- Room Assignment Task (this is "CSOP" in our data)
- Divergent Association Task (this is "DAT" in our data)
- Estimating Factual Quantities (this is "Becker" in our data)

An important initial piece of engineering work is to pull the data from the Task Map, and describe each task using its vector of features, rather than as a dummy variable that encodes the task name. This way, we can perhaps get at some of the "underlying differences" for tasks, rather than compare coarse bins to one another.

_Level 2: Add in Task-Specific Information_
In addition, some of our currently-ignored columns provide additional information / context for specific tasks. For example, CSOP problems vary in difficulty, and one of the ignored columns expresses the difficulty level of the task. In a later update, we should incorporate this information into the model --- although, since not every task has extra information, we will want to think about how to format the data.

### Adding More Inference Options / Modeling Tasks Better
Here, it's useful to add a bit of domain-knowledge. One common, if not somewhat oversimplified, model of tasks and the group process is that tasks are an input variable, which affect the process (for us, the communication) --> which in turn impact the output (performance).
![IPO Model](https://static.wixstatic.com/media/8faddc_abd87fa4f2c04f8eac18facff7afb909~mv2.jpg/v1/fill/w_600,h_363,al_c,q_80,usm_0.66_1.00_0.01,enc_auto/Untitled.jpg)'

This means that Tasks operate indirectly on the group's output, while the communication process is the direct mediator. This relationship isn't currently accounted for: we simply group together all features, and let the model do the work. 

Additionally, though the "Level 0" use of task features as dummy variables surfaces task variables as predictors, currently, it does nothing more than encode some statistical patterns in the underlying data. For example, this Shapley value simply tells me that, relative to other tasks, people had a lower mean performance on CSOP:
![Screenshot 2023-05-23 at 11 18 51 AM](https://github.com/Watts-Lab/team-process-map/assets/28793641/e631142e-044e-4cc8-9453-b155bae3b894)

Can we explore ways to explicitly add in the structure we know from domain knowledge, and use it to predict interesting relationships alongside conversation features?

Here are a few of my preliminary ideas/questions:

- **Explore the Task --> Process relationship more**: Is there a way that we can use the task to predict the communication process? As we had seen previously, it's possible to condense all of the communication processes relatively well into a PC dimension (although we should check if this is still the case after we fixed the lexical features). Can we predict how people will communicate just from the task? Currently, though, we have ~4-5 tasks and many, many conversations. So, perhaps we have to address this question in reverse: Can we use conversation features to classify people's conversations into bins?
- **Representing the variables as two distinct groups**: Can we add some sort of underlying structure such that tasks and conversation features are treated as two groups? Perhaps there can be a way to make the model hierarchical, so that one layer feeds forward into the next?
- **Interaction Effects**: is there some way to model 'interaction effects' (as in the traditional sense) between conversation and task features?

## Improving Usability
To improve usability, we should add the following (❗️ indicates most important):

- For Cross-Validation, make it possible for the user to specify a k (number of folds)
- ❗️Instead of defaulting to CV, create an option for the user to create a train-test split, train on X% of the data, and test on Y% of the data.
- Ensure train/test splits are balanced (E.g., mix of positive and negative examples)
- ⏩️ Allow the user to choose between multiple model options, not just XGBoost [To get us started on this, I added LASSO as one additional option.]
- ⏩️ Add visualization/metrics specific to each model option.
- Make it possible for the user to choose whether to use the raw or standardized target in the model.

## Improving Model Optimization
It seems that `optimize_model` almost always returns a model that is "worse" (e.g., worse R^2, MSE, etc.) than the baseline model. We may explore some of the following ways to improve the optimization:

-  Use a more efficient stopping criterion than 15 hardcoded trials (it is possible we are not exploring enough)
- Modify the hyperparameters being searched (perhaps we aren't searching the right space?)
- Move hyperparameter search to a GridSearch in the cloud

**Comment by yashveersinghsohi** [comment]
Updates and next steps as of 06/05.

The goal was to use the PR #112 as a way to deal with model training and the following 3 work streams: 

**Better Cross Validation**
- [x] Replace the `k-fold cross validation` implementation with a `train-val-test` split instead.
- [x] Allow a user to select the test set of their choice in the same pipeline.

**Model Performance**
- [ ] Use a better measure of performance for each dataset. For most datasets, selecting one column is not going to be a good proxy for the target we are trying to model (performance) and hence model performance will always be an issue
- [ ] Improve the hyperparameter tuning approach.

**Task Related Features and Feature Bagging**
- [x] Add in task related features as a way to distinguish tasks instead of one hot encoded task names.
- [ ] Understand a logical grouping/substructure within the features which can help us bag features more efficiently. This is turn will allow us to build better ensembles.

****

The ones checked are done in the PR linked above. The suggestion is to consider splitting the remaining tasks into their own separate branches as for now the branch `yashveer/baseline_model` does in-fact implement the "baseline" model only. 

In the future we could have the following branches to pursue the remaining tasks, and all the branches can be linked to this central issue:

# 1. `feature_bagging_eda`
This branch will go over the features from the datasets we have featurized up until now, and explore the features that show the most correlation with each other, and from a theoretical standpoint, compute the same values (i.e. they are trying to address the same issue - politeness, positivity, etc.) 

The idea behind this is to understand whether it would be wise to throw every single feature into the same model, or should we instead have different models focusing on different traits of conversations and aggregate their predictions in a meaningful way. 

Allowing for both of these approaches in the same pipeline will make the entire implementation over-flexible and very prone to bugs (we are allowing flexibility in dataset choices, target choices, model choices, optimization choices, and now possibly even model architecture choices!)

# 2. `performance_proxy_exploration`
This is by far the most ambiguous and the most critical problem in the model pipeline. Without reliable targets, we can not hope to build highly predictive models. We would have to come up with some notion of what performance looks like for each of the datasets we have which is mathematically intuitive as well. 

# 3. `improving_parameter_tuning`
Currently we are running bayesian optimization as a way to search the grid of hyperparameters automatically, and come up with a reasonable setting in the least amount of trials. We ideally would want to come up with some kind of `human in the loop` approach where we deduce the model performance characteristics of the baseline model (is it overfitting or under-fitting) based on better model diagnostics visuals (better and more informative SHAP scores etc.), and then choose the parameter space that makes the most sense to explore.

**Comment by xehu** [comment]
Closing as this is about a previous version of the project and no longer related to the toolkit
